diff --git a/dlls/ntdll/loadorder.c b/dlls/ntdll/loadorder.c
index 401d256..8a0b139 100644
--- a/dlls/ntdll/loadorder.c
+++ b/dlls/ntdll/loadorder.c
@@ -308,7 +308,7 @@ static HANDLE open_user_reg_key(const WCHAR *key_name)
     attr.SecurityQualityOfService = NULL;
     RtlInitUnicodeString( &nameW, key_name );
 
-    if (NtOpenKey( &hkey, KEY_ALL_ACCESS, &attr )) hkey = 0;
+    if (NtOpenKey( &hkey, KEY_WOW64_64KEY | KEY_ALL_ACCESS, &attr )) hkey = 0;
     NtClose( root );
 
     return hkey;
