diff --git a/Makefile b/Makefile
index 4c78883..8fec0c2 100644
--- a/Makefile
+++ b/Makefile
@@ -28,9 +28,9 @@ LDFLAGS += -Wl,--as-needed
 EXE  = nestopia
 LIBS = -lstdc++ -lm -lz -lasound $(shell sdl-config --libs) $(shell pkg-config --libs gtk+-3.0)
 
-PREFIX = /usr/local
-BINDIR = $(PREFIX)/bin
-DATADIR = $(PREFIX)/share/nestopia
+PREFIX = %GAMES_PREFIX%
+BINDIR = %GAMES_BINDIR%
+DATADIR = %GAMES_DATADIR%/nestopia
 
 # OpenGL Support
 CPPFLAGS += -DINCLUDE_OPENGL
@@ -183,14 +183,16 @@ $(EXE): $(OBJS)
 	$(CC) $(CFLAGS) $(LDFLAGS) -o $(EXE) $^ $(LIBS)
 
 install:
-	mkdir -p $(DATADIR)/icons
-	install -m 0755 $(EXE) $(BINDIR)
-	install -m 0644 source/linux/icons/nestopia.desktop $(DATADIR)
-	install -m 0644 NstDatabase.xml $(DATADIR)
-	install -m 0644 source/linux/icons/*.png $(DATADIR)/icons
-	install -m 0644 source/linux/icons/*.svg $(DATADIR)/icons
-	install -m 0644 source/linux/icons/nestopia.svg /usr/share/pixmaps
-	xdg-desktop-menu install --novendor $(DATADIR)/nestopia.desktop
+	mkdir -p $(DESTDIR)$(DATADIR)/icons
+	mkdir -p $(DESTDIR)/usr/share/applications
+	mkdir -p $(DESTDIR)/usr/share/pixmaps
+	mkdir -p $(DESTDIR)%GAMES_BINDIR%
+	install -m 0755 $(EXE) $(DESTDIR)$(BINDIR)
+	install -m 0644 source/linux/icons/nestopia.desktop $(DESTDIR)/usr/share/applications
+	install -m 0644 NstDatabase.xml $(DESTDIR)$(DATADIR)
+	install -m 0644 source/linux/icons/*.png $(DESTDIR)$(DATADIR)/icons
+	install -m 0644 source/linux/icons/*.svg $(DESTDIR)$(DATADIR)/icons
+	install -m 0644 source/linux/icons/nestopia.svg $(DESTDIR)/usr/share/pixmaps
 
 uninstall:
 	xdg-desktop-menu uninstall $(DATADIR)/nestopia.desktop
