diff -uNr irrlicht-9999_orig/examples/01.HelloWorld/Makefile irrlicht-9999/examples/01.HelloWorld/Makefile
--- irrlicht-9999_orig/examples/01.HelloWorld/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/01.HelloWorld/Makefile	2011-07-28 16:39:40.877281375 +0700
@@ -7,9 +7,9 @@
 # List of source files, separated by spaces
 Sources := main.cpp
 # Path to Irrlicht directory, should contain include/ and lib/
-IrrlichtHome := ../..
+IrrlichtHome := /usr
 # Path for the executable. Note that Irrlicht.dll should usually also be there for win32 systems
-BinPath = ../../bin/$(SYSTEM)
+BinPath = ../../bin
 
 # general compiler settings (might need to be set when compiling the lib, too)
 # preprocessor flags, e.g. defines and include paths
@@ -24,7 +24,7 @@
 #no changes necessary below this line
 ####
 
-CPPFLAGS = -I$(IrrlichtHome)/include -I/usr/X11R6/include $(USERCPPFLAGS)
+CPPFLAGS = -I/usr/include/irrlicht $(USERCPPFLAGS)
 CXXFLAGS = $(USERCXXFLAGS)
 LDFLAGS = $(USERLDFLAGS)
 
@@ -32,7 +32,7 @@
 all: all_linux
 
 # target specific settings
-all_linux all_win32 static_win32: LDFLAGS += -L$(IrrlichtHome)/lib/$(SYSTEM) -lIrrlicht
+all_linux all_win32 static_win32: LDFLAGS += -lIrrlicht
 all_linux: LDFLAGS += -L/usr/X11R6/lib$(LIBSELECT) -lGL -lXxf86vm -lXext -lX11 -lXcursor
 all_linux clean_linux: SYSTEM=Linux
 all_win32 clean_win32 static_win32: SYSTEM=Win32-gcc
diff -uNr irrlicht-9999_orig/examples/02.Quake3Map/Makefile irrlicht-9999/examples/02.Quake3Map/Makefile
--- irrlicht-9999_orig/examples/02.Quake3Map/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/02.Quake3Map/Makefile	2011-07-28 16:39:40.881281367 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -25,7 +25,7 @@
 static_win32 all_win32 clean_win32: SYSTEM=Win32-gcc
 static_win32 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32 static_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/03.CustomSceneNode/Makefile irrlicht-9999/examples/03.CustomSceneNode/Makefile
--- irrlicht-9999_orig/examples/03.CustomSceneNode/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/03.CustomSceneNode/Makefile	2011-07-28 16:39:40.889281351 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -23,7 +23,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/04.Movement/Makefile irrlicht-9999/examples/04.Movement/Makefile
--- irrlicht-9999_orig/examples/04.Movement/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/04.Movement/Makefile	2011-07-28 16:39:40.894281341 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -17,14 +17,14 @@
 endif
 
 # target specific settings
-all_linux: LDFLAGS = -L/usr/X11R6/lib$(LIBSELECT) -L../../lib/Linux -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -lXcursor
+all_linux: LDFLAGS = -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -ljpeg -lpng
 all_linux clean_linux: SYSTEM=Linux
 all_win32: LDFLAGS = -L../../lib/Win32-gcc -lIrrlicht -lopengl32 -lm
 all_win32: CPPFLAGS += -D__GNUWIN32__ -D_WIN32 -DWIN32 -D_WINDOWS -D_MBCS -D_USRDLL
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/05.UserInterface/Makefile irrlicht-9999/examples/05.UserInterface/Makefile
--- irrlicht-9999_orig/examples/05.UserInterface/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/05.UserInterface/Makefile	2011-07-28 16:39:40.901281327 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -23,7 +23,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/06.2DGraphics/Makefile irrlicht-9999/examples/06.2DGraphics/Makefile
--- irrlicht-9999_orig/examples/06.2DGraphics/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/06.2DGraphics/Makefile	2011-07-28 16:39:40.907281315 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -17,13 +17,13 @@
 endif
 
 # target specific settings
-all_linux: LDFLAGS = -L/usr/X11R6/lib$(LIBSELECT) -L../../lib/Linux -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -lXcursor
+all_linux: LDFLAGS = -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -ljpeg -lpng
 all_linux clean_linux: SYSTEM=Linux
 all_win32: LDFLAGS = -L../../lib/Win32-gcc -lIrrlicht -lopengl32 -lm
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/07.Collision/Makefile irrlicht-9999/examples/07.Collision/Makefile
--- irrlicht-9999_orig/examples/07.Collision/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/07.Collision/Makefile	2011-07-28 16:39:40.909281311 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -23,7 +23,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/08.SpecialFX/Makefile irrlicht-9999/examples/08.SpecialFX/Makefile
--- irrlicht-9999_orig/examples/08.SpecialFX/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/08.SpecialFX/Makefile	2011-07-28 16:39:40.915281299 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -17,13 +17,13 @@
 endif
 
 # target specific settings
-all_linux: LDFLAGS = -L/usr/X11R6/lib$(LIBSELECT) -L../../lib/Linux -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -lXcursor
+all_linux: LDFLAGS = -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -ljpeg -lpng
 all_linux clean_linux: SYSTEM=Linux
 all_win32: LDFLAGS = -L../../lib/Win32-gcc -lIrrlicht -lopengl32 -lm
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/09.Meshviewer/Makefile irrlicht-9999/examples/09.Meshviewer/Makefile
--- irrlicht-9999_orig/examples/09.Meshviewer/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/09.Meshviewer/Makefile	2011-07-28 16:39:40.918281293 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -23,7 +23,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/10.Shaders/Makefile irrlicht-9999/examples/10.Shaders/Makefile
--- irrlicht-9999_orig/examples/10.Shaders/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/10.Shaders/Makefile	2011-07-28 16:39:40.922281285 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -23,7 +23,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/11.PerPixelLighting/Makefile irrlicht-9999/examples/11.PerPixelLighting/Makefile
--- irrlicht-9999_orig/examples/11.PerPixelLighting/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/11.PerPixelLighting/Makefile	2011-07-28 16:39:40.928281273 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -17,13 +17,13 @@
 endif
 
 # target specific settings
-all_linux: LDFLAGS = -L/usr/X11R6/lib$(LIBSELECT) -L../../lib/Linux -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -lXcursor
+all_linux: LDFLAGS = -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -ljpeg -lpng
 all_linux clean_linux: SYSTEM=Linux
 all_win32: LDFLAGS = -L../../lib/Win32-gcc -lIrrlicht -lopengl32 -lm
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/12.TerrainRendering/Makefile irrlicht-9999/examples/12.TerrainRendering/Makefile
--- irrlicht-9999_orig/examples/12.TerrainRendering/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/12.TerrainRendering/Makefile	2011-07-28 16:39:40.931281266 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -17,13 +17,13 @@
 endif
 
 # target specific settings
-all_linux: LDFLAGS = -L/usr/X11R6/lib$(LIBSELECT) -L../../lib/Linux -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -lXcursor
+all_linux: LDFLAGS = -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -ljpeg -lpng
 all_linux clean_linux: SYSTEM=Linux
 all_win32: LDFLAGS = -L../../lib/Win32-gcc -lIrrlicht -lopengl32 -lm
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/13.RenderToTexture/Makefile irrlicht-9999/examples/13.RenderToTexture/Makefile
--- irrlicht-9999_orig/examples/13.RenderToTexture/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/13.RenderToTexture/Makefile	2011-07-28 16:39:40.933281261 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -23,7 +23,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/14.Win32Window/Makefile irrlicht-9999/examples/14.Win32Window/Makefile
--- irrlicht-9999_orig/examples/14.Win32Window/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/14.Win32Window/Makefile	2011-07-28 16:39:40.935281257 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -17,13 +17,13 @@
 endif
 
 # target specific settings
-all_linux: LDFLAGS = -L/usr/X11R6/lib$(LIBSELECT) -L../../lib/Linux -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -lXcursor
+all_linux: LDFLAGS = -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -ljpeg -lpng
 all_linux clean_linux: SYSTEM=Linux
 all_win32: LDFLAGS = -L../../lib/Win32-gcc -lIrrlicht -lopengl32 -lm
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/15.LoadIrrFile/Makefile irrlicht-9999/examples/15.LoadIrrFile/Makefile
--- irrlicht-9999_orig/examples/15.LoadIrrFile/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/15.LoadIrrFile/Makefile	2011-07-28 16:39:40.938281251 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -23,7 +23,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/16.Quake3MapShader/Makefile irrlicht-9999/examples/16.Quake3MapShader/Makefile
--- irrlicht-9999_orig/examples/16.Quake3MapShader/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/16.Quake3MapShader/Makefile	2011-07-28 16:39:40.940281247 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -17,13 +17,13 @@
 endif
 
 # target specific settings
-all_linux: LDFLAGS = -L/usr/X11R6/lib$(LIBSELECT) -L../../lib/Linux -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -lXcursor
+all_linux: LDFLAGS = -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -ljpeg -lpng
 all_linux clean_linux: SYSTEM=Linux
 all_win32: LDFLAGS = -L../../lib/Win32-gcc -lIrrlicht -lopengl32 -lm
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/18.SplitScreen/Makefile irrlicht-9999/examples/18.SplitScreen/Makefile
--- irrlicht-9999_orig/examples/18.SplitScreen/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/18.SplitScreen/Makefile	2011-07-28 16:39:40.942281243 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -23,7 +23,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/19.MouseAndJoystick/Makefile irrlicht-9999/examples/19.MouseAndJoystick/Makefile
--- irrlicht-9999_orig/examples/19.MouseAndJoystick/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/19.MouseAndJoystick/Makefile	2011-07-28 16:39:40.945281237 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 #CXXFLAGS = -O3 -ffast-math
 CXXFLAGS = -g -Wall
 
@@ -24,7 +24,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/20.ManagedLights/Makefile irrlicht-9999/examples/20.ManagedLights/Makefile
--- irrlicht-9999_orig/examples/20.ManagedLights/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/20.ManagedLights/Makefile	2011-07-28 16:39:40.947281233 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 #CXXFLAGS = -O3 -ffast-math
 CXXFLAGS = -g -Wall
 
@@ -24,7 +24,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/21.Quake3Explorer/Makefile irrlicht-9999/examples/21.Quake3Explorer/Makefile
--- irrlicht-9999_orig/examples/21.Quake3Explorer/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/21.Quake3Explorer/Makefile	2011-07-28 16:39:40.949281229 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp sound.cpp q3factory.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 #CXXFLAGS = -O3 -ffast-math
 CXXFLAGS = -g -Wall
 
@@ -23,7 +23,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/22.MaterialViewer/Makefile irrlicht-9999/examples/22.MaterialViewer/Makefile
--- irrlicht-9999_orig/examples/22.MaterialViewer/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/22.MaterialViewer/Makefile	2011-07-28 16:39:40.953281221 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 #CXXFLAGS = -O3 -ffast-math
 CXXFLAGS = -g -Wall
 
@@ -23,7 +23,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/23.SMeshHandling/Makefile irrlicht-9999/examples/23.SMeshHandling/Makefile
--- irrlicht-9999_orig/examples/23.SMeshHandling/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/23.SMeshHandling/Makefile	2011-07-28 16:39:40.956281215 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -23,7 +23,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/24.CursorControl/Makefile irrlicht-9999/examples/24.CursorControl/Makefile
--- irrlicht-9999_orig/examples/24.CursorControl/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/24.CursorControl/Makefile	2011-07-28 16:39:40.958281211 +0700
@@ -7,9 +7,9 @@
 # List of source files, separated by spaces
 Sources := main.cpp
 # Path to Irrlicht directory, should contain include/ and lib/
-IrrlichtHome := ../..
+IrrlichtHome := /usr
 # Path for the executable. Note that Irrlicht.dll should usually also be there for win32 systems
-BinPath = ../../bin/$(SYSTEM)
+BinPath = ../../bin
 
 # general compiler settings (might need to be set when compiling the lib, too)
 # preprocessor flags, e.g. defines and include paths
@@ -24,7 +24,7 @@
 #no changes necessary below this line
 ####
 
-CPPFLAGS = -I$(IrrlichtHome)/include -I/usr/X11R6/include $(USERCPPFLAGS)
+CPPFLAGS = -I/usr/include/irrlicht $(USERCPPFLAGS)
 CXXFLAGS = $(USERCXXFLAGS)
 LDFLAGS = $(USERLDFLAGS)
 
@@ -32,7 +32,7 @@
 all: all_linux
 
 # target specific settings
-all_linux all_win32 static_win32: LDFLAGS += -L$(IrrlichtHome)/lib/$(SYSTEM) -lIrrlicht
+all_linux all_win32 static_win32: LDFLAGS += -lIrrlicht
 all_linux: LDFLAGS += -L/usr/X11R6/lib$(LIBSELECT) -lGL -lXxf86vm -lXext -lX11 -lXcursor
 all_linux clean_linux: SYSTEM=Linux
 all_win32 clean_win32 static_win32: SYSTEM=Win32-gcc
diff -uNr irrlicht-9999_orig/examples/25.XmlHandling/Makefile irrlicht-9999/examples/25.XmlHandling/Makefile
--- irrlicht-9999_orig/examples/25.XmlHandling/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/25.XmlHandling/Makefile	2011-07-28 16:39:40.961281205 +0700
@@ -7,9 +7,9 @@
 # List of source files, separated by spaces
 Sources := main.cpp
 # Path to Irrlicht directory, should contain include/ and lib/
-IrrlichtHome := ../..
+IrrlichtHome := /usr
 # Path for the executable. Note that Irrlicht.dll should usually also be there for win32 systems
-BinPath = ../../bin/$(SYSTEM)
+BinPath = ../../bin
 
 # general compiler settings (might need to be set when compiling the lib, too)
 # preprocessor flags, e.g. defines and include paths
@@ -24,7 +24,7 @@
 #no changes necessary below this line
 ####
 
-CPPFLAGS = -I$(IrrlichtHome)/include -I/usr/X11R6/include $(USERCPPFLAGS)
+CPPFLAGS = -I/usr/include/irrlicht $(USERCPPFLAGS)
 CXXFLAGS = $(USERCXXFLAGS)
 LDFLAGS = $(USERLDFLAGS)
 
@@ -32,7 +32,7 @@
 all: all_linux
 
 # target specific settings
-all_linux all_win32 static_win32: LDFLAGS += -L$(IrrlichtHome)/lib/$(SYSTEM) -lIrrlicht
+all_linux all_win32 static_win32: LDFLAGS += -lIrrlicht
 all_linux: LDFLAGS += -L/usr/X11R6/lib$(LIBSELECT) -lGL -lXxf86vm -lXext -lX11 -lXcursor
 all_linux clean_linux: SYSTEM=Linux
 all_win32 clean_win32 static_win32: SYSTEM=Win32-gcc
diff -uNr irrlicht-9999_orig/examples/26.OcclusionQuery/Makefile irrlicht-9999/examples/26.OcclusionQuery/Makefile
--- irrlicht-9999_orig/examples/26.OcclusionQuery/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/26.OcclusionQuery/Makefile	2011-07-28 16:39:40.963281201 +0700
@@ -5,7 +5,7 @@
 Sources = main.cpp
 
 # general compiler settings
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -O3 -ffast-math
 #CXXFLAGS = -g -Wall
 
@@ -24,7 +24,7 @@
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 all_linux all_win32:
 	$(warning Building...)
diff -uNr irrlicht-9999_orig/examples/Demo/Makefile irrlicht-9999/examples/Demo/Makefile
--- irrlicht-9999_orig/examples/Demo/Makefile	2011-07-28 16:10:29.000000000 +0700
+++ irrlicht-9999/examples/Demo/Makefile	2011-07-28 16:39:40.965281197 +0700
@@ -2,7 +2,7 @@
 Target = Demo
 Sources = CDemo.cpp CMainMenu.cpp main.cpp
 
-CPPFLAGS = -I../../include -I/usr/X11R6/include
+CPPFLAGS = -I/usr/include/irrlicht
 CXXFLAGS = -Wall -O3 -ffast-math
 
 ifeq ($(HOSTTYPE), x86_64)
@@ -13,7 +13,7 @@
 
 # target specific settings
 all_linux: SYSTEM=Linux
-all_linux: LDFLAGS = -L/usr/X11R6/lib$(LIBSELECT) -L../../lib/$(SYSTEM) -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -lXcursor
+all_linux: LDFLAGS = -lIrrlicht -lGL -lXxf86vm -lXext -lX11 -lpng -ljpeg
 
 all_win32 clean_win32: SYSTEM=Win32-gcc
 all_win32: LDFLAGS = -L../../lib/$(SYSTEM) -lIrrlicht -lopengl32 -lm
@@ -25,7 +25,7 @@
 
 all_win32 clean_win32: SUF=.exe
 # name of the binary - only valid for targets which set SYSTEM
-DESTPATH = ../../bin/$(SYSTEM)/$(Target)$(SUF)
+DESTPATH = ../../bin/$(Target)$(SUF)
 
 OBJ = $(Sources:.cpp=.o)
 
